stk/backend/triton_kernels.py

```diff
@@ -0,0 +1,374 @@
+            BLOCK_M: tl.constexpr, BLOCK_N: tl.constexpr, BLOCK_K: tl.constexpr,
+            BLOCK_SIZE: tl.constexpr, GROUP_M: tl.constexpr, ACC_TYPE: tl.constexpr,
+            ):
+
+    # matrix multiplication
+    pid_m = tl.program_id(0)
+    pid_n = tl.program_id(1)
+    
+    num_pid_m = tl.num_programs(0)
+    num_pid_n = tl.num_programs(1)
+    pid_n, pid_m = tl.swizzle2d(pid_n, pid_m, num_pid_n, num_pid_m, GROUP_M)
+    
+    start_inx = tl.load(offsets + pid_m)
+    end_inx = tl.load(offsets + pid_m + 1)
+
+    pid = pid_m * 2 + pid_n
+
+    # pointers to sparse matrix
+    rm =  tl.arange(0, BLOCK_M)
+    rak = tl.arange(0, BLOCK_K)
+    # rm = tl.max_contiguous(tl.multiple_of(rm % M, BLOCK_M), BLOCK_M)
+
+    A += (rm[:, None] * stride_am + rak[None, :] * stride_ak)
+
+    # pointers to dense matrix
+    rn = pid_n * BLOCK_N + tl.arange(0, BLOCK_N)
+    rbk = tl.arange(0, BLOCK_K)
+    # rn = tl.max_contiguous(tl.multiple_of(rn % N, BLOCK_N), BLOCK_N)
+    B += (rbk[:, None] * stride_bk + rn[None, :] * stride_bn)
+
+    # do matrix multiplication
+    acc = tl.zeros((BLOCK_M, BLOCK_N), dtype=ACC_TYPE)
+    nsub_blocks = tl.cdiv(BLOCK_SIZE, BLOCK_K)
+
+    BLOCK_ELEMENTS = BLOCK_SIZE * BLOCK_SIZE
+    ak_sub_incr = BLOCK_K * stride_ak
+    bk_sub_incr = BLOCK_K * stride_bk
+    bk_block_incr = BLOCK_SIZE * stride_bk
+
+    # ak_sub_incr = tl.multiple_of(ak_sub_incr, BLOCK_K)
+    # bk_sub_incr = tl.multiple_of(bk_sub_incr, BLOCK_K)
+    # bk_block_incr = tl.multiple_of(bk_block_incr, BLOCK_SIZE)
+
+    for k in range(nsub_blocks * (end_inx - start_inx)):
+        sub_block_inx = k % nsub_blocks
+        block_inx = k // nsub_blocks
+        # mask = sub_block_inx == 0
+
+        if trans_A:
+            ptr_A = A + tl.load(block_offsets_t + start_inx + block_inx) * BLOCK_ELEMENTS + sub_block_inx * ak_sub_incr
+        else:
+            ptr_A = A + (start_inx + block_inx) * BLOCK_ELEMENTS + sub_block_inx * ak_sub_incr
+
+        ptr_B = B + tl.load(column_indices + start_inx + block_inx) * bk_block_incr + sub_block_inx * bk_sub_incr
+
+        # tl.debug_barrier()
+        # prev_offset_b = offset_b
+        # tl.store(debug + pid * 8 + k, offset_b)
+
+        a = tl.load(ptr_A)
+        b = tl.load(ptr_B)
+        acc += tl.dot(a, b) 
+
+    acc = acc.to(C.dtype.element_ty)
+
+    cm = pid_m * BLOCK_M + tl.arange(0, BLOCK_M)
+    cn = pid_n * BLOCK_N + tl.arange(0, BLOCK_N)
+
+    # if trans_C:
+    #     C = C + (cn[:, None] * stride_cm + cm[None, :] * stride_cn)
+    # else:
+    C = C + (cm[:, None] * stride_cm + cn[None, :] * stride_cn)
+    tl.store(C, acc, mask=True)
+
+@triton.autotune(
+    configs=[
+        # basic configs for compute-bound matmuls
+        triton.Config({'BLOCK_M': 128, 'BLOCK_N': 128, 'BLOCK_K': 32, 'BLOCK_SIZE': 128}, num_stages=4, num_warps=4),
+    ],
+    key=['M', 'N', 'K'],
+)
+@triton.jit
+def _dds_kernel(A, B, C, M, N, K,
```

Comment:
I think we'd decided that we could implement DDS from DSD that supports transposition of "C" to avoid duplicating code?

Reply:
There are correctness issues when combining them.